miner/env_changes.go

```diff
@@ -0,0 +1,422 @@
+	}
+
+	var (
+		// extract constraints into convenient format
+		refundIdx      = make([]bool, len(sbundle.Body))
+		refundPercents = make([]int, len(sbundle.Body))
+	)
+	for _, el := range sbundle.Validity.Refund {
+		refundIdx[el.BodyIdx] = true
+		refundPercents[el.BodyIdx] = el.Percent
+	}
+
+	var (
+		totalProfit      *big.Int = new(big.Int)
+		refundableProfit *big.Int = new(big.Int)
+
+		coinbaseDelta  = new(big.Int)
+		coinbaseBefore *big.Int
+	)
+
+	// insert body and check it
+	for i, el := range sbundle.Body {
+		coinbaseDelta.Set(common.Big0)
+		coinbaseBefore = c.env.state.GetBalance(c.env.coinbase)
+
+		if el.Tx != nil {
+			receipt, _, err := c.commitTx(el.Tx, chData)
+			if err != nil {
+				return err
+			}
+			if receipt.Status != types.ReceiptStatusSuccessful && !el.CanRevert {
+				return errors.New("tx failed")
+			}
+		} else if el.Bundle != nil {
+			err := c.commitSBundle(el.Bundle, chData, key, algoConf)
+			if err != nil {
+				return err
+			}
+		} else {
+			return errors.New("invalid body element")
+		}
+
+		coinbaseDelta.Set(c.env.state.GetBalance(c.env.coinbase))
+		coinbaseDelta.Sub(coinbaseDelta, coinbaseBefore)
+
+		totalProfit.Add(totalProfit, coinbaseDelta)
+		if !refundIdx[i] {
+			refundableProfit.Add(refundableProfit, coinbaseDelta)
+		}
+	}
+
+	// enforce constraints
+	coinbaseDelta.Set(common.Big0)
+	coinbaseBefore = c.env.state.GetBalance(c.env.header.Coinbase)
+	for i, el := range refundPercents {
+		if !refundIdx[i] {
+			continue
+		}
+		refundConfig, err := types.GetRefundConfig(&sbundle.Body[i], c.env.signer)
+		if err != nil {
+			return err
+		}
+
+		maxPayoutCost := new(big.Int).Set(core.SbundlePayoutMaxCost)
+		maxPayoutCost.Mul(maxPayoutCost, big.NewInt(int64(len(refundConfig))))
+		maxPayoutCost.Mul(maxPayoutCost, c.env.header.BaseFee)
+
+		allocatedValue := common.PercentOf(refundableProfit, el)
+		allocatedValue.Sub(allocatedValue, maxPayoutCost)
+
+		if allocatedValue.Cmp(common.Big0) < 0 {
+			return fmt.Errorf("negative payout")
+		}
+
+		for _, refund := range refundConfig {
+			refundValue := common.PercentOf(allocatedValue, refund.Percent)
+			refundReceiver := refund.Address
+			rec, err := c.commitPayoutTx(refundValue, c.env.header.Coinbase, refundReceiver, core.SbundlePayoutMaxCostInt, key, chData)
+			if err != nil {
+				return err
+			}
+			if rec.Status != types.ReceiptStatusSuccessful {
+				return fmt.Errorf("refund tx failed")
+			}
+			log.Trace("Committed kickback", "payout", ethIntToFloat(allocatedValue), "receiver", refundReceiver)
+		}
+	}
+	coinbaseDelta.Set(c.env.state.GetBalance(c.env.header.Coinbase))
+	coinbaseDelta.Sub(coinbaseDelta, coinbaseBefore)
+	totalProfit.Add(totalProfit, coinbaseDelta)
+
+	if totalProfit.Cmp(common.Big0) < 0 {
+		return fmt.Errorf("negative profit")
+	}
+	return nil
+}
+
+// revert reverts all changes to the environment - every commit operation must be followed by a revert or apply operation
+func (c *envChanges) revert() error {
```

Comment:
Hm, if we have apply and rollback (maybe apply-discard?), what is the corresponding commit to this revert function?
Also, what's the difference between revert and rollback? Shouldnt revert also clear txs and receipts?

Reply:
discussed offline, added comment and renamed function, please un-resolve if there are still concerns